src/wp-includes/kses.php

```diff
@@ -981,7 +982,18 @@ function wp_kses_split( $content, $allowed_html, $allowed_protocols ) {
 	$pass_allowed_html      = $allowed_html;
 	$pass_allowed_protocols = $allowed_protocols;
 
-	return preg_replace_callback( '%(<!--.*?(-->|$))|(<[^>]*(>|$)|>)%', '_wp_kses_split_callback', $content );
+	$token_pattern = <<<REGEX
+~
+	(                         # Detect comments of various flavors before attempting to find tags.
+		(?:<!--.*?(--!?>|$))  # Well-formed HTML comments like `<!-- ... -->` and also with invalid `--!>` closer.
+		|
+		</[^a-zA-Z][^>]*>     # Closing tags with invalid tag names.
```

Comment:
Stupid question: could we somehow reuse the same part of the regular expression `</[^a-zA-Z][^>]*>` below?

Reply:
Not stupid, but while we can do that, it's valuable IMO to leave this separate, as the _closing tag with an invalid tag name_ has an incredibly simple and straightforward syntax pattern: the tag name starts with _anything_ that isn't `a` - `z` (ASCII case-insensitive) and extends until the first `>`.

So by leaving this separate we can be even more careful about what we're examining. For example, opening tags with invalid tag names _are not interpreted as invalid comments_. The following pattern just isn't very robust in detecting normal tags let alone invalid ones, so as long as the correct approach is simpler being separate, I think it's worthwhile not joining them.